File: CPP/ByTier/B/1718_B_Fibonacci_Strings.cpp
// Problem: CF 1718 B - Fibonacci Strings
// https://codeforces.com/contest/1718/problem/B

/*
Purpose: Determine if a string composed of given letter frequencies can be rearranged to form a Fibonacci string.
Approach:
- Precompute Fibonacci sequence up to a large enough limit (80 terms).
- For a given set of letter frequencies, check if their total sum equals some Fibonacci number.
- Then try to assign the largest remaining frequencies to blocks of sizes matching the Fibonacci sequence in reverse order.
- If all blocks can be filled accordingly without violating the Fibonacci structure, output "YES", otherwise "NO".

Algorithms/Techniques:
- Fibonacci sequence generation
- Greedy assignment of frequencies to Fibonacci block sizes
- Mathematical validation of rearrangement possibility

Time Complexity: O(k^2) per test case where k is the number of letters in the alphabet.
Space Complexity: O(1) - fixed-size arrays for Fibonacci sequence and other variables.
*/

#include <stdio.h>
#include <iosfwd>

using namespace std;
typedef long long ll;
ll f[80], sum, cur;
int n, a[105], num[105];

void sol() {
  scanf("%d", &n);
  int t = 0;
  cur = 0, sum = 0;
  for (int i = 1; i <= n; i++)
    scanf("%d", &a[i]), sum += a[i]; // Read letter counts and compute total

  // Find the smallest Fibonacci number >= sum
  for (t = 0;; t++) {
    cur += f[t];
    if (cur >= sum)
      break;
  }

  // If total sum doesn't match a Fibonacci number, answer is NO
  if (cur != sum) {
    puts("NO");
    return;
  }

  int lst = -1; // Last used letter position
  // Process from largest Fibonacci number down to smallest
  for (t; ~t; t--) {
    int mx = -1, pos = -1;
    // Find the largest available count that is not the previously used letter
    for (int i = 1; i <= n; i++)
      if (a[i] > mx && i != lst)
        mx = a[i], pos = i;
    
    // If the largest available count is less than required Fibonacci block size, return NO
    if (mx < f[t]) {
      puts("NO");
      return;
    }
    
    // Subtract the block size from the selected letter's count
    a[pos] -= f[t];
    lst = pos; // Mark this letter as used
  }
  
  puts("YES");
}

int main() {
  f[0] = f[1] = 1;
  for (int i = 2; i < 80; i++)
    f[i] = f[i - 1] + f[i - 2]; // Generate Fibonacci sequence up to 80 terms
  
  int tc;
  scanf("%d", &tc);
  while (tc--)
    sol();
}


// https://github.com/VaHiX/CodeForces/

File: Python/ByTier/B/1718_B_Fibonacci_Strings.py
# Problem: CF 1718 B - Fibonacci Strings
# https://codeforces.com/contest/1718/problem/B

"""
Algorithm: Semi-Fibonacci String Checker

Approach:
1. Generate Fibonacci sequence up to a large enough number (60 terms).
2. For each test case:
   - Determine if the total number of characters can form a Fibonacci sum.
   - Try to assign character frequencies to Fibonacci numbers in reverse order.
   - Check if we can distribute all characters according to Fibonacci block sizes.

Time Complexity: O(m * k * log k + t * k * log k)
Where m is the number of Fibonacci terms generated (60), k is the alphabet size, and t is the number of test cases.
Space Complexity: O(m + k)
Where m is the size of Fibonacci array and k is the size of character frequency array.
"""

m = 60
fib = [1, 1]
for i in range(m):
    fib.append(fib[-1] + fib[-2])  # Generate Fibonacci sequence


def slv():
    n = int(input())
    _a = list(map(int, input().split()))
    a = []
    for i in range(n):
        a.append([_a[i], i])  # Store character counts with their indices
    j = -1
    for i in range(m):
        if sum(_a) == sum(fib[0 : i + 1]):  # Check if total chars match a Fibonacci sum
            j = i
            break
    if j == -1:
        print("no")  # No valid Fibonacci sum found
    else:
        _lst = -1  # Tracks last assigned character
        pok = 1    # Flag to indicate success/failure
        for k in range(j, -1, -1):  # Process in reverse Fibonacci order
            a.sort()  # Sort by character count
            need = fib[k]
            for i in range(n - 1, -1, -1):  # Iterate from largest count to smallest
                if a[i][1] == _lst:
                    continue  # Skip if already used this character
                if a[i][0] >= need and need:
                    x = min(need, a[i][0])
                    a[i][0] -= x
                    need -= x
                    _lst = a[i][1]
            if need > 0:  # Not enough characters to satisfy current Fibonacci number
                pok = 0
                break
        print("YES" if pok else "NO")


t = int(input())
for _ in range(t):
    slv()


# https://github.com/VaHiX/CodeForces/